src/umzug.ts

```diff
@@ -483,6 +480,22 @@ export class Umzug<Ctx extends object = object> extends emittery<UmzugEvents<Ctx
 		});
 	}
 
+	private async getContext(): Promise<Ctx> {
+		const isPromise = (_ctx: object | Function | PromiseLike<unknown>) => {
+			return 'then' in _ctx && typeof _ctx.then === 'function';
+		};
+
+		let { context = {} as Ctx }: UmzugOptions<Ctx> = this.options;
+
+		if (isPromise(context)) {
+			context = await Promise.resolve(context as Promise<Ctx>);
+		} else if (typeof context === 'function') {
+			context = await (context as () => Promise<Ctx> | Ctx)();
+		}
+
+		return context;
```

Comment:
Sorry, I clearly haven't been explaining this very well. We do not need to check if the context is a promise. The fact that this function is `async` will ensure it's one anyway, and a the js engine will flatten promises for us. The implementation of this method can and should be much simpler and have less risk of gotchas.

```suggestion
		const { context = {} } = this.options;
		return typeof context === 'function' ? context() : context;
```

The tests you added are good - they should pass if you make this change. Assuming they do, I'm happy to merge.